### PR TITLE
feat(dpo):  preference annotation improvements

### PR DESCRIPTION
**Pull request by danielsanchezaran**
# feat: DPO preference annotation improvements

## Summary

- **Critical bug fix**: DPO training was corrupting the model with a ~10m trajectory offset due to an un-normalized anchor in the loss computation
- **Initial pose pruning**: automatically reject trajectory pairs where the stochastic trajectory diverges from the deterministic one at the first pose
- **GT is Best button**: allow annotators to record the ground truth as the winner directly from the GUI
- **Classifier guidance system**: physically-meaningful alternative to noise scaling for generating the stochastic trajectory, with collision avoidance, lane keeping, and centerline following
- **Minor fixes**: time slider position preserved after regenerate; preference_collection.py return value unpack bug

---

### Bug fix: DPO loss normalization mismatch (trajectory offset)

In `dpo_loss.py::compute_trajectory_loss`, `current_states` (the anchor at index 0 of `all_gt`) was built from raw un-normalized `data["ego_current_state"][:, :4]`. In normal training the ego state is already normalized before the loss is called, placing the ego at `[-0.5, 0, 1, 0]` in normalized space. In DPO the raw value is `[0, 0, 1, 0]` — a 0.5-unit mismatch in normalized x corresponding to **10 metres** in real space (`normalization.json`: ego x-mean=10, std=20). The DPO gradient was therefore computed against a trajectory anchored 10m ahead of the actual vehicle position, causing the model to learn systematically offset trajectories.

**Fix:** normalize `ego_current` and `neighbors_current` with `ego_mean`/`ego_std` (already computed in the same function for `gt_trajectory_norm`) before building `current_states`.

---

### Initial pose pruning

When generating preference pairs the stochastic trajectory sometimes had a large position or heading offset relative to the deterministic one at the very first timestep, making the pair meaningless for annotation.

- New utilities in `preference_optimization/utils.py`:
  - `calculate_initial_position_displacement()` — Euclidean distance between first poses
  - `calculate_initial_yaw_difference()` — absolute yaw difference in degrees, correctly wrapped to [0, 180]
  - `should_prune_by_initial_pose()` — OR logic: flag the pair if position > threshold **or** yaw > threshold
- `generate_trajectory_pair()` extended with `enable_initial_pruning`, `initial_pos_threshold` (default 0.055 m), `initial_yaw_threshold_deg` (default 0.55°). The retry loop skips flagged candidates; if all retries fail it returns the last generated pair with `is_pruned=True`
- **Gradio UI:** master enable checkbox (default ON), position threshold slider (0.01–0.1 m), yaw threshold slider (0.1–1.0°)
- Visual behaviour:
  - Pruning **ON** + flagged → stochastic trajectory shown in **red**, selection button disabled
  - Pruning **OFF** + flagged → stochastic trajectory shown in **grey**, button stays enabled (visual indicator only, user can still annotate)
- 16 unit tests in `preference_optimization/test_initial_pruning.py`
- Side-fix: `preference_collection.py` was unpacking only 4 of 5 return values from `generate_trajectory_pair`

---

### Bug fix: time slider position not preserved after regenerate

`load_sample` and `regenerate` had `default_time = 40` hardcoded, causing the visualization to snap back to the middle timestep whenever the user regenerated or navigated, ignoring the time slider's current position. Fixed by threading `time_step` through all annotator methods and including `time_slider` in every Gradio event handler's inputs.

---

### GT is Best button

Adds a `🎯 GT is Best` button alongside the existing "Orange is Better" button, allowing annotators to record the ground truth trajectory as the winner when neither generated trajectory is acceptable.

- `_check_gt_available()` — validates `ego_agent_future` has ≥80% non-zero points
- `_get_gt_trajectory()` — converts NPZ GT `[T, 3]` (x, y, heading) to model format `[T, 4]` (x, y, cos, sin). No smoothing is applied: investigation of both `parse_rosbag.py` and `data_converter.cpp` confirmed that `ego_agent_future` is stored as raw Autoware EKF output with no additional processing, consistent with how the base model sees GT during training
- `select_gt_as_winner()` — records `(gt_trajectory, deterministic_output)` as `(traj_w, traj_l)` and advances to the next sample
- Button automatically disabled when GT is unavailable for the current sample
- Button styled red (`variant="stop"`) to distinguish it from trajectory selection
- WebSocket server: `select_gt_as_winner` action and `gt_available` state field added to `AnnotationParams`

---

### Classifier guidance system

Provides a physically meaningful alternative to noise-based diversity for generating the stochastic trajectory. When guidance is enabled, the DPM-Solver starts from zeros (deterministic starting point) and the guidance signal steers the trajectory toward the desired behaviour. `traj_1` (the green baseline) is always generated **without** guidance.

**Architecture:**
- `GuidanceWrapper` made configurable via `use_collision`, `use_route_following`, `use_lane_keeping`, `use_centerline_following` flags instead of hardcoding collision only
- `guidance_scale` exposed on `Config` and `Decoder` (was hardcoded at 0.5 in `decoder.py`)
- `generate_trajectory_pair()` accepts `enable_guidance`, per-function flags, and `guidance_scale`. Temporarily swaps `decoder._guidance_fn` for the duration of the call. When guidance is active the retry loop is bypassed since the guided output is fully deterministic
- `model_utils.load_model()` accepts guidance configuration flags

**New guidance functions:**
- `lane_keeping.py` — penalises ego protrusion beyond the left/right lane boundaries using actual vehicle width from `ego_shape`; activates only when the vehicle is outside the boundary (relu cost)
- `centerline_following.py` — quadratic cost on lateral distance from the nearest lane centreline; fires continuously so the correction is stronger the further the vehicle is from centre

**Fixes to existing guidance:**
- `route_following.py` — reshape bug used hardcoded `FeatureDim=12` instead of actual `SEGMENT_POINT_DIM=33`
- `collision.py` — cos/sin normalisation division by zero when model receives all-zero trajectory input; fixed with `.clamp(min=1e-6)`
- `guidance_wrapper.py` — replaced hard `assert not isnan` with per-function NaN check that skips the offending function and logs a warning instead of crashing the GUI

**Gradio UI:** new "🧭 Guidance" section with a master enable checkbox, individual checkboxes for each guidance function (Collision Avoidance, Route Following, Lane Keeping, Centerline Following), and a guidance scale slider (0.0–5.0, default 0.5)

---

## Example screen

image

## Test plan

- [ ] Run `python3 preference_optimization/test_initial_pruning.py` — all 16 tests should pass
- [ ] Launch DPO annotation GUI (`--preference_mode gui`) and verify:
  - Initial pose pruning correctly flags/blocks/colours trajectories in both ON and OFF modes
  - Time slider position is preserved after regenerate and after navigating between samples
  - GT is Best button records the GT as winner and is disabled when GT is unavailable
  - Guidance controls produce visually distinct stochastic trajectories for each guidance type
- [ ] Run a short DPO training session and confirm output trajectories no longer have a ~10m forward offset from the base_link origin

### CODE REVIEW

File: diffusion_planner/diffusion_planner/model/guidance/centerline_following.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    B, P, T_plus1, _ = x.shape
+    T = T_plus1 - 1
+
+    mask = (t < 0.1) * (t > 0.005)
+    mask = mask.view(B, 1, 1, 1)
+    x = torch.where(mask, x, x.detach())
+
+    ego_pos = x[:, 0, 1:, :2]  # [B, T, 2]
+
+    lanes = inputs["lanes"]  # [B, 140, 20, 33]
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply diffusion_planner/diffusion_planner/model/guidance/centerline_following.py]: I had the same question.

File: diffusion_planner/diffusion_planner/model/guidance/centerline_following.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            Required keys:
+                ``lanes`` – [B, N_seg, N_pts, SEGMENT_POINT_DIM]
+
+    Returns:
+        [B] energy tensor (higher = closer to centerline).
+    """
+    B, P, T_plus1, _ = x.shape
+    T = T_plus1 - 1
+
+    mask = (t < 0.1) * (t > 0.005)
```
Comment (SakodaShintaro) [inline diffusion_planner/diffusion_planner/model/guidance/centerline_following.py]: Very narrow range?

**Review by SakodaShintaro — APPROVED** [review]
@danielsanchezaran 
Basically, I think there is no problem.

It might be better to react to the points of GitHub Copilot.

File: diffusion_planner/diffusion_planner/model/guidance/centerline_following.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply diffusion_planner/diffusion_planner/model/guidance/centerline_following.py]: fixed in [140765b](https://github.com/tier4/Diffusion-Planner/pull/16/commits/140765bfe95b8c7eb14e3e4a60a953f973c2ee13)
